Use intrusive_ptr in Storage; replace unique_ptr<Storage> with Storage

ezyang opened the pull request:
```
Use intrusive_ptr in Storage; replace unique_ptr<Storage> with Storage

This patch does two major changes:

- It replaces the use of Retainable in Storage with a new implementation
  based on intrusive_ptr.  This will be necessary because Caffe2 will
  be using this class to implement intrusive_ptrs, and we need to
  line these up for the merge.  One good thing about the new implementation is
  that the default copy/move constructors/assignment operators and destructor
  work automatically, instead of needing to be hardcoded into Storage/Tensor.

- It replaces all places where we returned std::unique_ptr<Storage> with
  Storage, collapsing an unnecessary double indirection that is no longer
  necessary now that we have correctly working copy/move constructors.

I didn't initially want to do step (2), but it was very important to
eliminate all bare uses of new Storage and new StorageImpl, and this making
the API change was the most straightforward way to do this.

HOW TO FIX YOUR CODE IN THE NEW API

- You no longer need to dereference the result of tensor.storage() to pass
  it to set.  So, instead of:

      x.set_(*y.storage());

  just write:

      x.set_(y.storage());

- If you were accessing methods on StorageImpl via the pImpl() method, you
  must use the dot operator to run pImpl().  Even better; just drop pImpl,
  we now have method forwarding.  So, instead of:

      storage->pImpl()->data();

  just do:

      storage->data();
      // storage.pImpl()->data() works too but is not as recommended

MISC CODE UPDATES

- retain, release, weak_retain, weak_release and weak_lock are now
  reimplemented using the "blessed API"

- nvcc OS X and general OS X portability improvements to intrusive_ptr

- A new comment in intrusive_ptr describing how stack allocated
  intrusive_ptr_targets work differently than heap allocated ones
  from c10::make_intrusive

CAVEAT EMPTOR

- THStorage_weakRetain used to work on strong pointers, but it NO LONGER
  works with intrusive_ptr.  You must reclaim the strong pointer into a
  real strong pointer, construct a weak pointer from it, and then release
  the strong and weak pointers.  See StorageSharing.cpp for an example.
```

gchanan commented on aten/src/ATen/Storage.h:
```diff
@@ -7,7 +7,7 @@ namespace at {
 public:
   Storage() = delete;
-  Storage(StorageImpl* storage_impl) : storage_impl_(storage_impl) {}
+  Storage(StorageImpl* storage_impl) : storage_impl_(c10::intrusive_ptr<StorageImpl>::reclaim(storage_impl)) {}
```
so, the documentation of reclaim says "the pointer passed in *must* have been created using intrusive_ptr::release()" -- how do you know this is true?

ezyang replied:
You don't: it's an invariant calling code has to uphold. Essentially, we have both raw non-owning pointers and raw "owning" pointers, and it's only valid to call reclaim on raw owning pointers; calling it on a non-owning pointer will result in the refcount being less than the true refcount. There is no dynamic check for this (until you double free...)

There is also some business with stack allocated objects.

There are two ways a `T*` (where T inherits from `intrusive_ptr_target`) can be constructed. First is the ordinary way: you `new T` or stack allocate it. In this case, the reference count of T (in `intrusive_ptr_target` is initialized to zero as a special "sentinel" value; it doesn't mean that the refcount is actually zero, just that the object isn't managed by `intrusive_ptr`. `reclaim()` checks if the refcount is zero, and fails an assert if it is.

Second is via `make_intrusive`; in this case, the refcount is set to one, and the object will be destructed when its refcount goes to zero. In this case, if you get a raw pointer from it using `release()`, you can turn it back into an intrusive pointer with reclaim, and the refcount will be non-zero. (The danger case is when you get a non-owning raw pointer out using `get()`)

ezyang commented on aten/src/ATen/StorageImpl.h:
```diff
@@ -135,6 +116,50 @@ struct AT_API StorageImpl : public Retainable {
+  // handled automatically.
+  //
+  // TODO: Eliminate as many uses of these functions as humanly possible
+  void retain() {
```
I agree we should give weirder names, and I think I'm going to try to kill some of these uses.

gchanan commented on aten/src/ATen/TensorImpl.cpp:
```diff
@@ -62,9 +62,8 @@ void Tensor::backward(
-  StorageImpl* storage_impl = storage->pImpl();
-  tensor = new THTensor(storage_impl);
+  Storage storage = type->storage(true);
+  tensor = new THTensor(storage.release());
```
this is going away anyway, but shouldn't you release after you pass in to the tensor?

gchanan replied:
I guess it technically doesn't matter, but it looks scary.

ezyang replied:
Actually, there is something really scary here which we really should fix. In this context, `release` means, "relinquish ownership of the pointer managed by this smart pointer, and return the raw pointer" (which is immediately reclaimed by the `THTensor` here). However, on StorageImpl, we have `_raw_release`, which means "decrement this refcount". Probably should just rename `_raw_retain` and `_raw_release` to `_raw_incref` and `_raw_decref`.

gchanan commented on aten/src/ATen/Utils.h:
```diff
@@ -26,17 +26,16 @@ namespace at {
-  if (at::detail::get_backend(expr->pImpl()) != backend) {
-    AT_ERROR("Expected object of backend ", backend, " but got backend ", at::detail::get_backend(expr->pImpl()),
+static inline const Storage& checked_storage(const Storage& expr, const char * name, int pos, Backend backend, ScalarType scalar_type) {
+  if (at::detail::get_backend(expr.get()) != backend) {
```
You don't need to fix this now -- but this should take in a device type, no?

ezyang replied:
yep

gchanan commented on aten/src/ATen/templates/TypeDerived.cpp:
```diff
@@ -72,36 +72,34 @@ std::unique_ptr<Storage> ${Type}::storage(size_t size, bool resizable) const {
       InefficientStdFunctionContext::makeDataPtr(data, deleter,
 #if ${isCUDA}
-      Device(kCUDA, getPointerDevice(data))
+        Device(kCUDA, getPointerDevice(data))
```
making conflicts with yourself? :).

gchanan commented on test/cpp/api/rnn.cpp:
```diff
@@ -110,15 +110,15 @@ TEST_CASE("rnn") {
     LSTM model(2, 2);
     for (auto& v : model->parameters()) {
       float size = v->numel();
-      auto p = static_cast<float*>(v->data().storage()->pImpl()->data());
```
why do we allow -> on storages?  There can't be that many useful methods, right?  Just move them to Storage directly?  Because we don't have the issue that we have with Tensor where we need a ton of methods here (i.e. TensorMethods.h).

ezyang replied:
Per offline conversation, "Just because I'm lazy." We can add the methods on them pretty easy later.

gchanan commented on aten/src/ATen/SparseTensorImpl.cpp:
```diff
@@ -62,10 +62,7 @@ TensorImpl* SparseTensorImpl::maybe_zero_dim(bool condition_when_zero_dim) {
-  AT_ERROR("sparse tensors do not have storage");
-}
-at::StorageImpl* SparseTensorImpl::storageImpl() const {
+const Storage& SparseTensorImpl::storage() {
```
I don't really buy that there's no downside here -- the user needs to understand how storages are ref-counted, otherwise they can hold on to a dangling references.  I guess it's fine at the Impl level where presumably people understand what is going on, but at the Tensor level it's not obvious to me that that is the case.

ezyang replied:
Offline convo:

Storage will never dangle as long as Tensor is live. Instead, you'd get the following behavior:

```
Storage s = x.storage(); // equivalent: auto s = ...
x.set_(y.storage());
// s still points to the original storage of x (before set_)

// versus

const Storage& s = x.storage();
x.set_(y.storage());
// s now points to y's storage
```

This is idiomatic C++ behavior, and the non-reference case (which is default) matches PyTorch behavior in Python:

```
>>> x = torch.zeros(2)
>>> y = torch.ones(2)
>>> s = x.storage()
>>> s
 0.0
 0.0
[torch.FloatStorage of size 2]
>>> x.set_(y.storage())
tensor([1., 1.])
>>> s
 0.0
 0.0
[torch.FloatStorage of size 2]
>>> x
tensor([1., 1.])
```

gchanan commented on aten/src/ATen/Storage.h:
```diff
@@ -19,29 +19,36 @@ struct AT_API Storage {
+  DeviceType device_type() const { return storage_impl_->device_type(); }
+  at::Allocator* allocator() const { return storage_impl_.get()->allocator(); }
+  at::ScalarType scalar_type() const { return storage_impl_->scalar_type(); }
+  int getDevice() const { return storage_impl_->getDevice(); }
```
Please no integers as devices.  Can we have a device / deviceType / whatever makes sense at this level? (I don't particularly care about the lower levels).

ezyang replied:
I'm going to replace this with `device` returning a `Device`.

gchanan commented on aten/src/ATen/Storage.h:
```diff
@@ -19,29 +19,36 @@ struct AT_API Storage {
+  int getDevice() const { return storage_impl_->getDevice(); }
+
+  StorageImpl* unsafeReleaseStorageImpl() {
+    return storage_impl_.release();
```
Why do we need this here over calling unsafeGetStorageImpl() followed by release at the call sites?

ezyang replied:
Offline agreement: get rid of this method, and rewrite the one call site to manually incref itself.

gchanan commented on aten/src/ATen/Storage.h:
```diff
@@ -6,8 +6,8 @@ namespace at {
 public:
-  Storage() = delete;
-  Storage(StorageImpl* storage_impl) : storage_impl_(storage_impl) {}
+  Storage() {}
```
I forget if I asked about this -- do we actually need this?

ezyang replied:
We don't have to give Storage a default constructor, but since it's a pointery thing, it seems more idiomatic C++ to give it one. Makes reasonable things work (like storing Storage in containers).

gchanan replied:
Yah it's just...then we get back to the UndefinedTensor question, which I don't really want to deal with :).  But it's up to you.

ezyang commented on aten/src/TH/THStorageFunctions.cpp:
```diff
@@ -57,20 +40,19 @@ ptrdiff_t THStorage_size(const THStorage *self)
     if (size != 0) {
-      storage->set_data_ptr(
-          storage->allocator()->allocate(storage->elementSize() * size));
+      new_data = storage->allocator()->allocate(storage->elementSize() * size);
```
Actually, we did; previously `storage->data_ptr()` returned a mutable reference, so `std::swap` on it resulted in us making the storage pointer null.

gchanan replied:
Oh, nice!

ezyang commented on aten/src/TH/THTensor.cpp:
```diff
@@ -196,3 +195,10 @@ THTensor_compute_stride(at::IntList oldshape, at::IntList oldstride, at::IntList
+// NB: Steals ownership of storage
+void THTensor_stealAndSetStoragePtr(THTensor* tensor, THStorage* storage) {
+  // Caffe2 might have tensors whose storages are null, but we
+  // don't allow it in PyTorch.
```
Yeah, this looks wrong, should have assert.

ezyang committed:
Use intrusive_ptr in Storage; replace unique_ptr<Storage> with Storage (#10488)

Summary:
```
Use intrusive_ptr in Storage; replace unique_ptr<Storage> with Storage

This patch does two major changes:

- It replaces the use of Retainable in Storage with a new implementation
  based on intrusive_ptr.  This will be necessary because Caffe2 will
  be using this class to implement intrusive_ptrs, and we need to
  line these up for the merge.  One good thing about the new implementation is
  that the default copy/move constructors/assignment operators and destructor
  work automatically, instead of needing to be hardcoded into Storage/Tensor.

- It replaces all places where we returned std::unique_ptr<Storage> with
  Storage, collapsing an unnecessary double indirection that is no longer
  necessary now that we have correctly working copy/move constructors.

I didn't initially want to do step (2), but it was very important to
eliminate all bare uses of new Storage and new StorageImpl, and this making
the API change was the most straightforward way to do this.

HOW TO FIX YOUR CODE IN THE NEW API

- You no longer need to dereference the result of tensor.storage() to pass
  it to set.  So, instead of:

      x.set_(*y.storage());

  just write:

      x.set_(y.storage());

- If you were accessing methods on StorageImpl via the pImpl() method, you
  must use the dot operator to run pImpl().  Even better; just drop pImpl,
  we now have method forwarding.  So, instead of:

      storage->pImpl()->data();

  just do:

      storage->data();
      // storage.pImpl()->data() works too but is not as recommended

- storage->getDevice() is no more; instead use storage->device().index()

MISC CODE UPDATES

- retain, release, weak_retain, weak_release and weak_lock are now
  reimplemented using the "blessed API", and renamed to make it
  clearer that their use is discouraged.

- nvcc OS X and general OS X portability improvements to intrusive_ptr

- A new comment in intrusive_ptr describing how stack allocated
  intrusive_ptr_targets work differently than heap allocated ones
  from c10::make_intrusive

CAVEAT EMPTOR

- THStorage_weakRetain used to work on strong pointers, but it NO LONGER
  works with intrusive_ptr.  You must reclaim the strong pointer into a
  real strong pointer, construct a weak pointer from it, and then release
  the strong and weak pointers.  See StorageSharing.cpp for an example.
```
Pull Request resolved: https://github.com/pytorch/pytorch/pull/10488

Differential Revision: D9306134

fbshipit-source-id: d242ab92d0dcf72c4b96e29fe933debf5ebf07cb

gchanan commented on aten/src/ATen/StorageImpl.h:
```diff
@@ -108,43 +88,111 @@ struct AT_API StorageImpl : public Retainable {
+    ptr.release();
+    return wptr.release();
+  }
+  void _raw_weak_incweakref() {
```
I think one weak is enough :).

ezyang replied:
It's actually an important distinction.

When I have a `StorageImpl*` there are two types of raw pointer it could be. It could be a *strong* pointer, which is only permissible when `refcount > 0`, or it could be a *weak* pointer, which is permissible for any value of refcount, even when `refcount == 0`. In the old, `Retainable` implementation, we didn't really distinguish between these two cases, which means you could try to increment the (strong) refcount, even when you only had a weak pointer. That would be invalid if `refcount == 0`; once you lose all strong references, you are never allowed to get a strong reference to the object again. But nothing checked for this.

In the new `c10::intrusive_ptr` implementation, the implementations of `intrusive_ptr::reclaim` and `weak_intrusive_ptr::reclaim` implementations have much more stringent asserts: `intrusive_ptr` asserts that there MUST have been a strong pointer which was released; similarly, `weak_intrusive_ptr` asserts that there MUST have been a weak pointer that was released. So, if you for example, call `_raw_weak_incweakref` on a strong pointer, and the object has no weak pointers on it, the assert will conclude that this is an invalid use of `weak_intrusive_ptr::reclaim`, because your `weakcount == 1` (and that weak reference is implied by the strong reference), so you couldn't have possibly release'd a weak pointer that you are now reclaiming, you must have done something wrong.

So, `_raw_weak_incweakref` is ONLY valid if you call it on a "weak" `StorageImpl*` pointer. That's what the first weak means. And it increments the weak reference count; thus the second weak.

gchanan replied:
Can there be a `_raw_incweakref`?  Or `_raw_weak_incref` / `raw_weak_decref`?

gchanan replied:
I guess my assumption was, as you mentioned, there are two regimes of refcounts: weak and strong.  Anyone using the raw interface would know this so say, "raw_weak_incref" would have a clear meaning.  But I can see the argument the other way as well, so do what you think is best!

gchanan reviewed (approved):
nice!
